### ISSUE TITLE
Let's finally implement selectable formatters and custom args

### ISSUE THREAD

**Issue by lassik** [lassik/emacs-format-all-the-code#27]
I've been procrastinating like crazy on the custom args support. I meant to do it already in early January. Please complain in this thread so I can have some social pressure to get it done :D

**Comment by lassik** [comment]
I'll also give direct write access to contributors. There was a great blog post somewhere about broad write access, development speed and motivation but I can't find it now. My own attention has been increasingly gravitating toward other things after the initial push to get package this into a usable state (my temperament is also heavily biased towards fast instead of steady development). It's in all our best interest if this project is developed by the people who are the most motivated to do it.

**Comment by lassik** [comment]
Also, for anyone who is excited about auto-formatting, [Unibeautify](https://github.com/unibeautify) could still use some more people :)

On the Emacs front, @purcell recently made [reformatter.el](https://github.com/purcell/reformatter.el) as an alternative approach exploring the same territory as format-all.

I've been toying with the idea of auto-generating formatter definitions from JSON. The same JSON could be used by all kinds of packages (Unibeautify, Emacs, Vi, etc.). The Unibeautify main authors are committed to implementing formatters as one npm module each.

Personally, my main interest would be to get as many people collaborating as we can so we can develop a common language for configuring formatters. The Unibeautify vision is closest to this.

**Comment by lassik** [comment]
@terlar @ztlevi and @knu now have write access, should they want it :)

We could make all changes via pull request from now on. (I will also use PRs instead of direct commits like everybody else.) In my experience, this is the clearest workflow for multi-contributor projects - it's easy for everyone to keep track of what's going on and do code reviews (often it really helps to have a second and third opinion).

We could just accept PRs by consensus. Everyone with write access is free to merge them. The project has been quite stagnant so it's probably better for the development process to prioritize moving faster instead of getting everything just right on the first try.

**Comment by lassik** [comment]
Here's what I had in mind for the multi-formatter stuff: it could be done with a single buffer-local variable, `format-all-formatter-list`, which lists one or more formatters to run.

    (defvar-local format-all-formatter-list t
      "Which formatter (or formatters) to use for the current buffer.
    
    The value is a list of formatters.  A formatter can be just a
    symbol FORMATTER-NAME.  It can also be a list (FORMATTER-NAME
    \"arg1\" \"arg2\" ... \"argN\") to add extra arguments to the end
    of the command line.  Note that the formatter name is always a
    symbol, whereas the command line arguments are always strings.
    By default, the formatter is not given any command line arguments
    that affect how the code will look, i.e. the formatter will use
    its default style or consult its configuration files.
    
    Normally you'll want each buffer to be formatted by one formatter
    only.  But it is possible to chain multiple formatters together
    in this list.  In that case they are all run, in the order given,
    every time the buffer is formatted.  If any of them fails, that's
    an error and the buffer is reverted to the way it was before
    formatting.  The point is that this is not a list of alternative
    formatters from which the first available one is chosen; rather,
    it's a chain where each formatter alters the code in some useful
    way and the result is a combination of them all.
    
    If the formatter list is empty (i.e. nil) the current buffer is
    not formatted at all, even when `format-all-buffer' is called.
    
    Finally, instead of a list the value may be the symbol t.  Then
    format-all uses a default formatter based on the `major-mode' of
    the buffer with no extra command line arguments.  This is the
    global default in case you don't set any buffer-local value.
    
    In addition to Emacs Lisp hooks, a formatter can also be set
    interactively via C-u M-x `format-all-buffer'.
    
    Examples:
    
        (setq format-all-formatter-list t)
        (setq format-all-formatter-list '(clang-format))
        (setq format-all-formatter-list '((black \"-S\")))")

* I think this is slightly complex for the most common case (just run a single formatter with the default arguments) but maybe not too bad, considering even the most complex cases could be handled with this same variable.

* Should we continue to have default formatters? Seems good for usability, but potentially error-prone for scripting.

* I've been intending to refactor the formatter definitions to use `[language-id.el](https://github.com/lassik/emacs-language-id/blob/master/language-id.el)` to detect the language. A couple of the current definitions (esp. prettier) have got so messy that I have to concentrate to make sense of them every time I look at them :)

**Comment by rassie** [comment]
Hmm. I don't think I like the buffer-local approach very much, since I want a solution where I can keep using Emacs while my other team members keep using whatever they want and we still stay on the same page. My current problem is the python formatter: `format-all-the-code` uses `black` by default, while e.g. VS Code is probably using something from LSP, but definitely not `black`. So the concept of `unibeautify` is fairly spot-on for me: a common platform, a project-specific configuration and clear commit history about which tools are used, so that each team member can adapt their environment to the project. Using buffer-local variables is probably too much clutter for non-emacs users to be bothered by.

**Comment by lassik** [comment]
Thanks for chiming in.

The point of format-all is just to offer a selection of formatters - it's completely up to you to decide whether and when to use them. Unibeautify is also based on the same principle. But it has a formatter configuration file that is checked into Git, which is a social signal that others programmers on the team should also be using Unibeautify with the same settings.

Much of the point of format-all is to have your own personal hook function (usually in `after-change-major-mode-hook`) to enable or disable it for your own needs. So you would configure it to match whatever your team expects, and it you can't match those expectations, just disable it in those situations. This is far from perfect. The point of this issue is to add selectable formatters and args so your hook function could be more flexible and cover more situations, which would bring things closer to perfection, but it's still a bit of a painful way to configure things compared to something like Unibeautify (which we both agree would be the right thing).

The buffer-local formatter/args approach would allow you to do everything you are doing now - it would strictly increase flexibility so you can use more formatters and more of their options. The clutter would be in your personal `init.el` file so your teammates wouldn't ever see it. (You could have `.dir-locals.el` files in projects but at that point I would just think about contributing to Unibeautify to get it developed faster 😎)

**Comment by mpereira** [comment]
Hi @lassik,

First of all, thanks for writing this package.

Custom executables and args is something that interests me. Do you think a mapping from modes to "executable & args" would be sufficiently flexible to abide to most people's needs?

I was thinking something along the lines of:
```elisp
(setq format-all-formatter-alist
      '((sh-mode (:executable "shfmt"
                  :args ("-ln"
                         "-i" "2")))
        (yaml-mode (:executable "prettier"
                    :args ("--parser" "yaml")))))
```

And regarding

> Should we continue to have default formatters?

IMO yes.

**Comment by lassik** [comment]
Thanks for the comments.

I think specifying the formatter on a per-mode basis would not be fine-grained enough. Lots of people work on many projects simultaneously, and different projects have different coding styles for the same language. I would go for something like `(directory mode formatter args)`, letting Emacs automatically pick the right mode as it usually does and then pick the formatter based on its decision.

**Comment by mpereira** [comment]
> Lots of people work on many projects simultaneously, and different projects have different coding styles for the same language

For this specific criteria, would [directory-local variables](https://www.gnu.org/software/emacs/manual/html_node/emacs/Directory-Variables.html) be a good solution?

https://endlessparentheses.com/a-quick-guide-to-directory-local-variables.html

**Comment by lassik** [comment]
Yes. Basically, all we need to do is define a buffer-local variable (e.g. `format-all-formatter-list` as above) that says which formatter(s) to use and what (if any) extra command line args to give. Users can set that variable from directory-local variables, from hooks, or from anywhere else they prefer.

**Comment by ianpan870102** [comment]
> ...from directory-local variables, from hooks, or from anywhere else they prefer

I second this idea, this seems to be able to solve most people's needs. On top of that, I think it's reasonable to **use mode-hooks as the first thing users set, then directory-local-variables is set with higher priority**. The motivation behind this idea is for e.g. For all personal Python projects I want to use **black**, but only one specific project I want to use **autopep8** because that's what my team went with. Then we would be able to hook general python-mode to formatting with **black**, and add this particular directory to be **autopep8** formatted. Since you intended to support chaining, perhaps it would be nice to have the option to disable a formatter, so that one could choose between chaining **autopep8** after **black**, or directly **override black** with **autopep8** for a specific project.

> Should we continue to have default formatters?

Yes, definitely IMO!

**Comment by lassik** [comment]
I'll finally implement this.

Here's the plan:

* Use the [language-id](https://github.com/lassik/emacs-language-id) library to figure out the GitHub Linguist language ID for the current buffer.
* Select the right formatter based on the language ID. (Currently the formatter is selected per mode. Under the new arrangement, the mode->langauge detection is done by the language-id library and format-all doesn't need to know about it.)

Then:

* We can have more than one formatter for one language-id.
* Each language-id has a default formatter (which will be the same formatters that are used now).
* In `.dir-locals.el` you can set the buffer-local variable `format-all-formatters` to specify formatters (and extra flags for each formatter) per language. For example:

```
((format-all-formatters
  ("C" clang-format)
  ("SQL" sqlformat "--indent_width" "4")
  ("JavaScript" standard)))
```

In principle, it's also possible to write `-*- format-all-formatters: ... -*-` or `Local Variables:` directly into a source file, but since Emacs and format-all are quite niche, I would discourage such comments and recommend either having a separate `.dir-locals.el` or using personal hooks in each person's init file.

Let me know if there are problems with this arrangement.

**Comment by lassik** [comment]
We should probably also make it so that if `format-all-formatters` is defined (non-nil) yet doesn't say anything about the language of the current buffer, then `format-all-buffer` skips formatting the current buffer (i.e. skips using the default formatter even if there is a default).

**Comment by lassik** [comment]
The syntax to chain multiple formatters for one language should also be figured out. It could be as simple as:

```
((format-all-formatters
  ("Python" autopep8 "--flags-for-autopep8" "..." black "--flags-for-black" "...")))
```

So each symbol in the list gives the name of a new formatter, and any strings following that symbol (before the next symbol or the end of the list) are extra command line args for that formatter.

**Comment by mpereira** [comment]
Great to see movement here, I appreciate your time @lassik!

Your proposal sounds good. Fwiw I like the second alternative better.

**Comment by lassik** [comment]
I'm also starting to prefer the second alternative, seems a bit cleaner. Let's go with it.

**Comment by terlar** [comment]
Will it still also support using the buffer file name in different places to the command? As some formatters require the filename in order to format (for example eslint for cache and and figuring out what format rules to apply based on file extension)?

**Comment by lassik** [comment]
Formatting files by filename causes so much trouble that it's not worth it IMHO. Since all the formatters are open source, it's better to submit a feature request or PR to the few formatters that don't yet support stdin->stdout formatting. That's much better design and benefits all editor plugins, not just format-all.

Some formatters have a `--stdin-filename` flag that can be used to say, "pretend that the contents of stdin come from a file with this name". It's used for things like where to look for configuration files and what filename to show in error messages. I imagine it could also be used for caching.

**Comment by lassik** [comment]
To be clear, format-all has never supported formatting by filename, but it does pass buffer-file-name to those formatters that have a `--stdin-filename` flag, so from a cursory inspection it looks like it does.

**Comment by lassik** [comment]
Inserting `buffer-file-name` into custom command line arguments is something that we could do, but it sounds like that kind of thing would be better handled by format-all's own formatter definitions. The user-given command-line arguments are intended to be added to the ones that format-all currently gives. The idea is that all the complicated plumbing (what filenames and file descriptors to pass, how to report errors, etc.) is handled by format-all and users can just pass the flags that affect the coding style for the formatted text.

If the formatter definitions can be made more useful for everyone by passing extra flags, we can always add those flags there. What do you think of this approach?

**Comment by lassik** [comment]
In fact eslint already has a stdin-filename flag 😄 https://eslint.org/docs/user-guide/command-line-interface#stdin-filename

**Comment by ianpan870102** [comment]
Great to see this feature getting implemented! I really appreciate this project @lassik , this is always among the first packages I install when I move my emacs config to a new computer.

**Comment by webframp** [comment]
Looking forward to this feature, thanks @lassik for the work on it!

**Comment by ianpan870102** [comment]
Does anybody know if there are any updates regarding this thread?

**Comment by lassik** [comment]
Just more procrastination. Sorry about that. Need to rewrite the internals to rely on `language-id`.

**Comment by lassik** [comment]
Finally got around to doing PR #78. It switches format-all's language detection to rely on the external language-id library. This makes the internals much cleaner, and was the biggest problem in the way of multi-formatter support.

**Comment by lassik** [comment]
Merged that PR to master. It should be in MELPA in a few hours, and the user-visible behavior in the new version of format-all should be the same as the old version. Please report if anything broke so we can fix those problems before adding multi-formatter.

If you do `(setq format-all-debug t)` you can watch the language detection in action. The `*Messages*` buffer now says something like: `Format-All: Formatting format-all.el using (emacs-lisp "Emacs Lisp")`. The `"Emacs Lisp"` is the language name from the `language-id` library.

The multi-formatter thing would work so you can use these language IDs in dir-locals and hooks to specify which formatter you want.

**Comment by lassik** [comment]
I've never added a cross-dependency from one MELPA library to another before, so hopefully nothing breaks. The lib is here: http://melpa.org/#/language-id

**Comment by ianpan870102** [comment]
> The multi-formatter thing would work so you can use these language IDs in dir-locals and hooks to specify which formatter you want.

Could you please give use an example to use hooks to specify formatters? For instance, if I want to use prettier instead of tidy for HTML modes. Cheers in advance

**Comment by lassik** [comment]
@ianpan870102 Thanks! The changes 4 days ago didn't yet add multi-formatter/custom args support, they were just a big refactoring. Nobody has complained that the refactoring broke anything, so here the features finally are.

Since some issues will probably come up, I pushed them into a `multi-formatter` branch instead of directly pushing to `master` and hence to MELPA.

People interested in multi-formatter and/or custom args support: please test the branch and report anything that is broken or confusing in this thread.

There's a new variable `format-all-formatters` that can be set in a hook function or in `.dir-locals.el`. It's buffer-local, but the most convenient thing is probably to set it for all your buffers globally, and/or override it per-project in a `.dir-locals.el` file.

`standard` has been added as another choice for JavaScript (besides `prettier`), and Lua now has Prettier's lua-plugin in addition to lua-fmt.

**Comment by lassik** [comment]
Also, `M-x format-all-buffer` now shows a minibuffer prompt in case there are multiple formatters available for the current buffer's language and you haven't selected one yet. `C-u M-x format-all-buffer` forces the prompt to appear even if you already selected a formatter before. That way you can change it if you made the wrong choice.

`format-all-buffer` doesn't save this formatter choice to any file, so it's reset the if you close and re-open the file. To preserve the choice, set `format-all-formatters` manually in a hook or in `.dir-locals.el`. The minibuffer prompt also doesn't let you set custom args or chain more than one formatter into a pipeline. Setting `format-all-formatters` lets you do both of those if you want to. Configuring things from the minibuffer every time you open a file is quite clumsy so I put only the simplest behavior (selecting one formatter and using its default options) in there.

**Comment by lassik** [comment]
Also `format-all-mode` no longer does any formatting unless `format-all-formatters` has a formatter for the current buffer's language. That means you either have to run `M-x format-all-buffer` to set it, or configure `format-all-formatters` in a hook or in `.dir-locals.el`.

This is preparation to change `format-all-mode` from a buffer-local minor mode into a global minor mode (so you can have it always on like font-lock if you want to). For now, it's still a buffer-local mode.

**Comment by lassik** [comment]
Here's an example `.dir-locals.el` file:

```lisp
((nil
  (format-all-formatters
   ("JavaScript" standard)
   ("Python" (black "--skip-string-normalization"))
   ("SQL" (sqlformat "-a" "--keywords" "capitalize")))))
```

The `nil` means that the variable is set for all Emacs modes. Obviously you can combine the format-all settings with other settings you'd have in a dir-locals file. You can also set `format-all-settings` only for particular modes, though I'm not sure there's any particular benefit to that.

**Comment by lassik** [comment]
File comments also work: `// -*- format-all-formatters: (("JavaScript" standard)) -*-`

Though your non-Emacs teammates will probably not appreciate it :-) I'd recommend going with a `.dir-locals.el` at the root of the Git repo.

**Comment by ianpan870102** [comment]
@lassik Hi! Has there been updates regarding merging the branch for multiple formatters into MELPA? I noticed in #82 you planned to do it a few months back. No rush :) Just curious to see how things have been.

**Comment by lassik** [comment]
@ianpan870102 Thanks for the continued interest. No updates - I've just been slacking off. I've been trying to recruit co-maintainers for a couple of years but nobody has volunteered for the job so development has been slow.

The `multi-formatter` branch worked fine last time I checked, but I couldn't make a decision about whether to turn `format-all-mode` into a global mode or not. Multi-formatter support introduces a buffer-local variable saying what formatters (if any) to use for each buffer. This variable can be set to a nil value to disable formatting for a particular buffer. Hence global mode would make sense. What do you think?

**Comment by ianpan870102** [comment]
Hi @lassik Lassi, thanks for getting back. I've always been interested in this project, albeit my emacs-lisp skills don't allow me to contribute much. I really appreciate your work!

I've tried out the multi-formatter branch this week but Emacs keeps prompting me about safe local variables. Even if I chose to allow, I can't seem to get it to detect my `.dir-locals.el`. 

At the moment, I'm trying to use `yapf` instead of `black` for Python formatting, and the only solution that I've found so far is to manually paste this block into `format-all.el` and byte compile it:

```emacs-lisp
(define-format-all-formatter yapf
  (:executable "yapf")
  (:install "pip install yapf")
  (:languages "Python")
  (:format (format-all--buffer-easy executable)))
```

This "hack" works on the master branch. But I want to get your multi-formatter approach to work. I don't mind setting a `.dir-locals.el` file in my directory or even modify `format-all-formatters` variable directly in my `init.el`. However, I tried many combinations but still can't get it to work...

For instance, I tried to include a super simple `.dir-locals.el` like so:

```emacs-lisp
((nil
  (format-all-formatters
   ("Python" yapf))))
```

or set `format-all-formatters` to `'(("Python" yapf))`, but both didn't work out.



> I couldn't make a decision about whether to turn format-all-mode into a global mode or not.

IMO it's cleaner to make it a minor-mode and let users add-hooks to their prog-modes. Personally I prefer interactively calling `format-buffer` manually anyways, as I prefer to not have my editor auto-format on save.

**Comment by lassik** [comment]
Sorry about the delay. I installed `yapf`, tried the exact things you did (i.e. add the `.dir-locals.el` with those contents and add the `yapf` formatter definition) and it works fine for me.

**Comment by lassik** [comment]
> IMO it's cleaner to make it a minor-mode and let users add-hooks to their prog-modes. Personally I prefer interactively calling `format-buffer` manually anyways, as I prefer to not have my editor auto-format on save.

OK, let's keep it as a buffer-local mode for now on. Thanks for giving your opinion :) Designing in a one-person echo chamber is not optimal.

**Comment by lassik** [comment]
I just committed the yapf formatter definition you gave into the `multi-formatter` branch in commit. Commit `93329ce`.

This works fine on my computer. For debug messages, try `M-x eval-expression (setq format-all-debug t)` and look at the end of the `*Messages*` buffer after running `M-x format-all-buffer`.

This is what I get:

```
Format-All: Formatting formatexample.py as Python using yapf
Format-All: Running: /usr/local/bin/yapf
Format-All: Directory: /home/lassi/tmp/format-all/
Reformatted!
```

**Comment by ianpan870102** [comment]
Thanks for getting back @lassik , I just tried the `.dir-locals.el` approach again and it worked (yapf formats my python buffer) -- turns out last time I **forgot to byte-compile** my `format-all.el` after replacing the master branch with the `multi-formatter` branch.

If I may, I suggest that we still keep the default formatters around, and give local variables (user-specified formatters) a higher priority. I'm saying this because formatting some other languages' buffers (e.g. JavaScript, C++) now yields "**No formatter**".

**Comment by lassik** [comment]
Glad that you got it to work!

You're right, in this branch there is no default formatter anymore - for each buffer, you have to first do `M-x format-all-buffer` interactively to pick one (or set `format-all-formatters` in a Lisp hook or `.dir-locals.el` file). This also means that `format-all-mode` (the format-on-save hook) does not do anything unless you have chosen a formatter first.

**Comment by lassik** [comment]
The `format-all-mode` situation should really be rectified somehow. I keep forgetting to run `format-all-buffer` first and then wonder why saving doesn't do anything.

**Comment by lassik** [comment]
> If I may, I suggest that we still keep the default formatters around, and give local variables (user-specified formatters) a higher priority. I'm saying this because formatting some other languages' buffers (e.g. JavaScript, C++) now yields "**No formatter**".

That error only happens when you run `(format-all-buffer)` non-interactively. What kind of setup do you get it with? If you do it interactively via `M-x`, it should ask you for `Formatter for JavaScript:` in the minibuffer the first time.

**Comment by ianpan870102** [comment]
> That error only happens when you run (format-all-buffer) non-interactively. What kind of setup do you get it with? If you do it interactively via M-x, it should ask you for Formatter for JavaScript: in the minibuffer the first time.

Ah, that's probably coz I'm calling a custom function that in turn calls your `format-all-buffer`. I'm doing it like this:

```emacs-lisp
(defun my/format-code ()
  "Auto-format whole buffer."
  (interactive)
  (if (derived-mode-p 'prolog-mode)
      (prolog-indent-buffer)
    (format-all-buffer)))
```

**Comment by lassik** [comment]
`(call-interactively 'format-all-buffer)` might solve your immediate problem. But we should really find a better way to do it on format-all's side.

**Comment by sgleizes** [comment]
@lassik Hi! Thanks for this much needed package.
I see a lot of great improvements here and it seems like this is close to a merge. What do you think should be done before merging into MELPA? Restoring default-formatters maybe?

**Comment by lassik** [comment]
OK, the multi-formatter work has finally been merged into the `master` branch! The WIP `multi-formatter` branch is gone.

Let me know if anything breaks.

Format-all now supports:

* Adding custom args to command-line formatters
* Running more than one formatter for the same buffer in a chain

`format-all-formatters` is the key variable that controls everything. The variable is buffer-local. The best place to set it is probably `.dir-locals.el`. In that way, you can set the formatters for an entire project in one place.

The default value for `format-all-formatters` is the empty list (i.e. `nil`) which means no formatters are run by default.

When you run `M-x format-all-buffer` it finds the formatter chain that pertains to the current buffer's language in `format-all-formatters`. If no such chain is found, it asks you for a formatter to use in the minibuffer. (If format-all supports only one formatter for the buffer's language, it doesn't ask, and simply uses that formatter, just as it has done so far.)

`C-u M-x format-all-buffer` (i.e. `format-all-buffer` with a prefix argument) always prompts you which formatter to use, overriding and changing the value of `format-all-formatters` for the buffer.

`format-all-mode` is still a buffer-local minor mode, not a global minor mode. If `format-all-formatters` doesn't specify a formatter for the language in the current buffer, it doesn't format anything. It doesn't ask you which formatter to use or set a default formatter like the `M-x format-all-buffer` command does. This behavior is conservative in an attempt not to break anything; it's debatable that the mode should set a formatter or at least show a warning that none has been set.

**Comment by lassik** [comment]
Here's an example `.dir-locals.el` file:

```Emacs Lisp
((nil
  (format-all-formatters
   ("JavaScript" standard)
   ("Python" (yapf "--style" "google")))))
```

**Comment by ianpan870102** [comment]
Quick question, now that I have a `.dir-locals.el` file in my project directory, Emacs will prompt me with "The local variables ... contains values that may not be safe ... Do you want to apply it?"
What's the recommended way to deal with this? I don't wish to be prompted every time I open a buffer -- should I type `!` and mark them as safe whenever I visit a new file? How do people normally workaround this?

**Comment by lassik** [comment]
Yes, pressing `!` is the recommended thing to do. Then Emacs will remember your choice and won't ask you each time.

Note that if you change anything in the `.dir-locals.el` file,

**Comment by lassik** [comment]
The feature is now finally in master and this issue has received no comments in the last 2 weeks, so I'm closing it. Feel free to open new issues if there are problems or feature requests.

**Comment by ianpan870102** [comment]
Is it possible (as an option) to automatically run function "format-all-ensure-formatter" when visiting a buffer? I almost always want to use the default formatter and only occasionally want to use a self-defined `.dir-locals.el`.
I guess what I'm trying to say is, keep the default formatter enabled (just like in the older versions of format-all package) but give `.dir.locals.el` a higher priority, if such file exists.
Let me know what you think! Just throwing out ideas here.
Happy New Year!

**Comment by lassik** [comment]
Happy belated new year to you as well!

`(add-hook 'format-all-mode-hook 'format-all-ensure-formatter)` should do what you want. I just checked that if you have both that hook **and** formatters defined in `.dir-locals.el`, the `.dir-locals.el` formatters will prevail, so the hook will do what you expect whether a file has dir-locals or not.

**Comment by lassik** [comment]
The reason `format-all-mode` doesn't automatically ensure a default formatter is that I've been keeping the option open of turning it into a global mode later on. If it's global, people will want some buffers not to be formatted. The mode is currently buffer-local.

**Comment by ianpan870102** [comment]
Thanks @lassik! Works like a charm.
Just wanna add on: for those that do not need automatic formatting before save, and thus don't have format-all-mode enabled, the following line works in a similar manner:

```emacs-lisp
(add-hook 'prog-mode-hook 'format-all-ensure-formatter)
```

**Comment by ltylty** [comment]
in doom emacs , do it like this 
```
(set-formatter! 'my-sqlformat "sqlformat -r -")
(setq-hook! 'sql-mode-hook +format-with 'my-sqlformat)
```
